qiskit/transpiler/passes/layout/vf2_layout.py

```diff
@@ -145,10 +144,21 @@ def run(self, dag):
         # Filter qubits without any supported operations. If they don't support any operations
         # They're not valid for layout selection
         if self.target is not None:
-            has_operations = set(itertools.chain.from_iterable(self.target.qargs))
-            to_remove = set(range(len(cm_nodes))).difference(has_operations)
-            if to_remove:
-                cm_graph.remove_nodes_from([cm_nodes[i] for i in to_remove])
+            has_none = False
+            has_operations = set()
+            for qarg in self.target.qargs:
+                if qarg is None:
+                    has_none = True
+                else:
+                    has_operations.update(qarg)
+            # If there are defined operations or if has_operations is empty and there are
+            # no globally defined gates. If has_none was true and there are no operations
+            # defined that means all operations are supported on all qubits and there is
+            # nothing to remove.
+            if has_operations or not has_none:
+                to_remove = set(range(len(cm_nodes))).difference(has_operations)
+                if to_remove:
+                    cm_graph.remove_nodes_from([cm_nodes[i] for i in to_remove])
```

Comment:
```suggestion
            has_operations = {q for p in self.target.qargs for q in p}
            to_remove = set(range(len(cm_nodes))).difference(has_operations)
            if to_remove:
                cm_graph.remove_nodes_from(cm_nodes[to_remove])
```

Reply:
A comment to replace what I removed might be useful.
*There are similar changes to make below. I will do this tomorrow*

Reply:
As we discussed offline last month the proposed set of changes here isn't quite right. The logic here is a bit nuanced based on some intricacies in the `Target` class's data model when dealing with globally available ideal gates. Basically the target for any operation can typically defines a dictionary mapping qargs to `InstructionProperties` objects which represents a real operation on the qargs. There is an alternative mode of operation when qargs is `None` this is used to indicate an "ideal" operation on all valid qargs for the backend. The original intent for using `None`  was for ideal simulator backends which don't have any constraints on qubits or any properties for the transpiler to track.

When we have backends with a mixed set of real/noisy instructions in the target and some ideal instructions we need to be careful in how we handle it. The desired behavior here is that if there are any real instructions we restrict the output based on those and basically just add the ideal globally available instructions on top of those connectivity constraints. So this block:

```
            # If there are defined operations or if has_operations is empty and there are
            # no globally defined gates. If has_none was true and there are no operations
            # defined that means all operations are supported on all qubits and there is
            # nothing to remove.
            if has_operations or not has_none:
```

is guarding against the case when we don't have any defined real/noisy operations (i.e. they're all ideal) if we didn't have `not has_none` here `to_remove` would be the set of all instructions and we'd remove all the qubits from the coupling graph which is the opposite of the desired effect (although we probably wouldn't ever reach this point because the coupling graph should be `None` and this pass doesn't have any work to do). The suggested changes won't work because by moving the `None in qargs` check to the outer condition in the case of mixed real and ideal instructions in the `Target` we won't do the filtering based on any idle qubits, based on the defined real/noisy instructions.

That being said I'll take another look at this to see if there is some simplification to logic I can make here to simplify the logical flow to make it easier to follow, because I've looked at this code twice now since I originally wrote it back in August and it does take me a little bit of time to remind myself of how the logic works.